### PR TITLE
Improve the Android SDK rolling process

### PR DESCRIPTION
**Pull request by dnfield**
This allows us to download the Android SDK andNDK directly from dl.google.com, rather than the previous process which involves manually setting up the SDK locally, building a tarball, and uploading it to a mojo bucket.

This will require changes in the engine, which I'll prepare naother PR for to preview.

### CODE REVIEW

**Comment by dnfield** [comment]
See also https://github.com/flutter/flutter/issues/17780

File: tools/android/android_sdk_downloader/lib/src/zip.dart [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  await outDir.parent.create(recursive: true);
+  final Directory tempDir = await outDir.parent.createTemp();
+
+  String command;
+  List<String> args;
+  if (Platform.isWindows) {
+    command = 'powershell.exe  -nologo -noprofile -command '
+        '"& { '
+        'Add-Type -A \'System.IO.Compression.FileSystem\'; '
+        '[IO.Compression.ZipFile]::ExtractToDirectory(\'$file\', \'${tempDir.path}\'); '
```
Comment (goderbauer) [inline tools/android/android_sdk_downloader/lib/src/zip.dart]: note: this mechanism of extracting things is kinda super-slow. I know, it's also the only one that available on pretty much all Windows platforms, so...

File: tools/android/android_sdk_downloader/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+Flutter engine development.
+
+## Usage
+
+```
+-r, --repository-xml            Specifies the location of the Android Repository XML file.
+                                (defaults to "https://dl.google.com/android/repository/repository2-1.xml")
+
+-p, --platform                  Specifies the Android platform version, e.g. 28
+                                [28 (default), 22]
```
[the remark this answers is not in the document]
Comment (dnfield) [reply tools/android/android_sdk_downloader/README.md]: Stale code - will update

**Comment by chinmaygarde** [comment]
Is this good to go?

File: tools/android/android_sdk_downloader/lib/src/android_repository.dart [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+  final String text;
+}
+
+class AndroidRepositoryRemotePackage {
+  const AndroidRepositoryRemotePackage(
+    this.revision,
+    this.displayName,
+    this.archives, {
+    this.isObsolete = false,
+  })  : assert(revision != null),
```
Comment (cbracken) [inline tools/android/android_sdk_downloader/lib/src/android_repository.dart]: nit: Are the two spaces here intentional? I believe our style is just one.

File: tools/android/android_sdk_downloader/lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        help: 'Automatically accept Android SDK licenses.')
+    ..addFlag(
+      'overwrite',
+      defaultsTo: false,
+      help: 'Skip download if the target directory exists.',
+    );
+
+  final bool help = args.contains('-h') ||
+      args.contains('--help') ||
+      (args.isNotEmpty && args.first == 'help');
```
Comment (cbracken) [inline tools/android/android_sdk_downloader/lib/main.dart]: Flutter style: boolean operators on the left of the newlines.

File: tools/android/android_sdk_downloader/lib/main.dart [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  if (help) {
+    print(argParser.usage);
+    return;
+  }
+
+  final Options options = Options.parseAndValidate(args, argParser);
+
+  final AndroidRepository androidRepository =
+      await _getAndroidRepository(options.repositoryXmlUri);
```
Comment (cbracken) [inline tools/android/android_sdk_downloader/lib/main.dart]: Flutter style: Un-linebreak. (here and a bunch of other places)

File: tools/android/android_sdk_downloader/lib/src/android_repository.dart [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+// found in the LICENSE file.
+
+import 'package:xml/xml.dart' as xml show parse;
+import 'package:xml/xml.dart';
+
+// see https://android.googlesource.com/platform/tools/base/+/master/sdklib/src/main/java/com/android/sdklib/repository/sdk-repository-10.xsd
+
+const String _xsi = 'http://www.w3.org/2001/XMLSchema-instance';
+const String _sdk =
+    'http://schemas.android.com/sdk/android/repo/repository2/01';
```
Comment (cbracken) [inline tools/android/android_sdk_downloader/lib/src/android_repository.dart]: Flutter style: begin global consts with a `k` or `_k`.

File: tools/android/android_sdk_downloader/lib/main.dart [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+      'repository-xml',
+      abbr: 'r',
+      help: 'Specifies the location of the Android Repository XML file.',
+      defaultsTo: _androidRepositoryXml,
+    )
+    ..addOption(
+      'platform',
+      abbr: 'p',
+      help: 'Specifies the Android platform version, e.g. 28',
+      defaultsTo: '28',
```
[the remark this answers is not in the document]
Comment (dnfield) [reply tools/android/android_sdk_downloader/lib/main.dart]: This was really in there for testing.  I have no issue taking it out -in the actual consuming script I'm expecting that we'd pass these explicitly.

File: tools/android/android_sdk_downloader/lib/src/http.dart [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+class DownloadTracker {
+  DownloadTracker(this.name, this.total) : received = 0;
+
+  final String name;
+  final int total;
+  int received;
+
+  String get percent =>
+      '${((received / total) * 100).round().toString().padLeft(3)}%';
```
[the remark this answers is not in the document]
Comment (dnfield) [reply tools/android/android_sdk_downloader/lib/src/http.dart]: yeah this is weird either way but I'll gow ith the shorter one.

File: tools/android/android_sdk_downloader/lib/src/options.dart [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+class OptionsRevision {
+  const OptionsRevision(
+    this.raw, [
+    this.major = 0,
+    this.minor = 0,
+    this.micro = 0,
+    this.preview = 0,
+  ]);
+
+  factory OptionsRevision.fromRaw(String raw) {
```
[the remark this answers is not in the document]
Comment (dnfield) [reply tools/android/android_sdk_downloader/lib/src/options.dart]: Done

**Pull request merged**